File: lab/source-tree/vulnerable-code/vulnerable.js
// Semgrep: Command injection
const { exec } = require('child_process');
function vulnerable_exec(userInput) {
  exec(`echo ${userInput}`, (error, stdout, stderr) => {});  // VULNERABLE
}

// Semgrep: Hardcoded secrets
const API_SECRET = 'super_secret_key_12345';
const DB_PASS = 'mysql_root_password';

// Semgrep: SQL injection
function vulnerable_query(userId) {
  return `SELECT * FROM users WHERE id = ${userId}`;  // VULNERABLE
}

// Semgrep: Eval usage
function dangerous_eval(code) {
  eval(code);  // VULNERABLE
}

// Semgrep: XXE in XML parsing
const xml2js = require('xml2js');
const parser = new xml2js.Parser({ 
  doctype: true  // VULNERABLE: Allows XXE
});

// Semgrep: No input validation
function process_input(data) {
  return data.toUpperCase();  // No sanitization
}

File: lab/source-tree/vulnerable-code/vulnerable.py
import os
import subprocess

# Semgrep: OS injection vulnerability
def vulnerable_command_execution(user_input):
    os.system(f"echo {user_input}")  # VULNERABLE: OS injection

# Semgrep: Hardcoded credentials
DATABASE_PASSWORD = "admin123password"
API_KEY = "sk-12345abcde67890fghij"

# Semgrep: SQL injection
def vulnerable_query(user_id):
    query = f"SELECT * FROM users WHERE id = {user_id}"
    return query

# Semgrep: Path traversal
def read_file(filename):
    with open(f"/files/{filename}", "r") as f:
        return f.read()

# Semgrep: Weak cryptography
import hashlib
def hash_password(password):
    return hashlib.md5(password.encode()).hexdigest()  # VULNERABLE: MD5

# Semgrep: Insecure deserialization
import pickle
def deserialize_user(data):
    return pickle.loads(data)  # VULNERABLE: Unsafe pickle
